packages/util/src/liquidityProxy.ts

```diff
@@ -0,0 +1,1207 @@
+    ],
+  };
+};
+
+// output token is xor, user indicates desired output amount
+// x - other token reserve
+// y - xor reserve
+// y_out - desired output amount (xor)
+const xykQuoteD = (x: FPNumber, y: FPNumber, yOut: FPNumber): QuoteResult => {
+  const y1 = safeDivide(yOut, ONE.sub(XYK_FEE));
+
+  if (FPNumber.isGreaterThanOrEqualTo(y1, y)) {
+    throw new Error(`xykQuote: output amount ${y1.toString()} is larger than reserves ${y.toString()}. `);
+  }
+
+  const xIn = safeDivide(x.mul(y1), y.sub(y1));
+
+  return {
+    amount: xIn,
+    fee: y1.sub(yOut),
+    rewards: [],
+    distribution: [
+      {
+        market: LiquiditySourceTypes.XYKPool,
+        amount: yOut,
+      },
+    ],
+  };
+};
+
+const xykQuote = (
+  inputReserves: FPNumber,
+  outputReserves: FPNumber,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  isXorInput: boolean
+): QuoteResult => {
+  try {
+    if (isDesiredInput) {
+      if (isXorInput) {
+        return xykQuoteA(inputReserves, outputReserves, amount);
+      } else {
+        return xykQuoteB(inputReserves, outputReserves, amount);
+      }
+    } else {
+      if (isXorInput) {
+        return xykQuoteC(inputReserves, outputReserves, amount);
+      } else {
+        return xykQuoteD(inputReserves, outputReserves, amount);
+      }
+    }
+  } catch (error) {
+    return safeQuoteResult(amount, LiquiditySourceTypes.XYKPool);
+  }
+};
+
+// AGGREGATOR
+const quotePrimaryMarket = (
+  inputAssetId: string,
+  outputAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  payload: QuotePayload
+): QuotePrimaryMarketResult => {
+  if ([inputAssetId, outputAssetId].includes(XSTUSD)) {
+    return {
+      result: xstQuote(inputAssetId, amount, isDesiredInput, payload),
+      market: LiquiditySourceTypes.XSTPool,
+    };
+  } else {
+    return {
+      result: tbcQuote(inputAssetId, outputAssetId, amount, isDesiredInput, payload),
+      market: LiquiditySourceTypes.MulticollateralBondingCurvePool,
+    };
+  }
+};
+
+// xor is output
+const primaryMarketAmountBuyingXor = (
+  collateralAssetId: string,
+  amount: FPNumber,
+  isDesiredInput: boolean,
+  xorReserve: FPNumber,
+  otherReserve: FPNumber,
+  payload: QuotePayload
+): FPNumber => {
+  try {
+    const secondaryPrice = FPNumber.isGreaterThan(xorReserve, FPNumber.ZERO)
+      ? safeDivide(otherReserve, xorReserve)
+      : MAX;
+
+    const primaryBuyPrice =
+      collateralAssetId === XSTUSD
+        ? xstBuyPriceNoVolume(collateralAssetId, payload)
+        : tbcBuyPriceNoVolume(collateralAssetId, payload);
+
+    const k = xorReserve.mul(otherReserve);
+
+    if (isDesiredInput) {
+      if (FPNumber.isLessThan(secondaryPrice, primaryBuyPrice)) {
+        const amountSecondary = k.mul(primaryBuyPrice).sqrt().sub(otherReserve);
+
+        if (FPNumber.isGreaterThanOrEqualTo(amountSecondary, amount)) {
+          return FPNumber.ZERO;
+        } else if (FPNumber.isLessThanOrEqualTo(amountSecondary, FPNumber.ZERO)) {
```

Comment:
please add a function inside this file for `FPNumber.isLessThanOrEqualTo(something, FPNumber.ZERO)`